Automattic/jetpack
# Boost: Fix files linting formatting and integrate into pre-commit hooks and CI workflow.

Fixes #21066

The overall purpose of this PR is to make sure that all the JavaScript, TypeScript and Svelte Boost files are correctly linted/reformated as well as provide the correct commands and setup to integrate linting/reformating from the command line in the Boost directory as well as in the pre-commit git hooks and in the CI.

#### Changes proposed in this Pull Request:

This PR does:

* Fix the `projects/plugins/boost/.eslintrc.js` file so the `js, ts, svelte` files can all be linted properly according to their parsers and rules.
* Fix the `js, ts, svelte` files in the Boost directory according to the valid rules now properly triggered by the fixed `.eslintrc.js` file from the boost directory.
* Remove the `projects/plugins/boost/.eslintignore` file so all the ignore rules are now inside the `.eslintignore` at the root of the Monorepo. 
* Cleanup the `.eslintignore` at the root of the Monorepo so it removes obsolete rules as well re-enable linting for some js files in the root of the Boost folder as well as for the e2e tests files. Also, I have added the `!.prettierrc.js` statement as the `.prettierrc.js` changed but without that line, I was not able to commit due to an error message mentioning that this file was already excluded by default and could fix the problem by using the negative exclude pattern.
* Remove the `projects/plugins/boost/.prettierrc.js` file and move all its content (minus the `require` statement for `@wordpress/prettier-config` which does not seem needed) in the `/.prettierrc.js` at the root of the Monorepo. This was the only way that I found to have all the `js, ts, svelte` Boost files to be properly formatted by prettier. I have also moved the prettier package requirement and upgraded prettier into the `tools/js-tool/package.json` file. It was easier that way.
*  Cleanup the `projects/plugins/boost/package.json` removing obsolete packages, adding a couple (so eslint in PHPStorm can work properly for Boost) and clean up the script commands to ensure we are also linting e2e tests with the lint command.
* Update the `tools/js-tools/git-hooks/pre-commit-hook.js` to ensure that `svelte` files can be linted/formatted during the pre-commit git hook.
* Update the `github/workflows/linting.yml` workflow to ensure that `svelte` files can be linted during the CI workflow. Also, add the step to install the Jetpack Boost Test e2e JS node packages during the linting process so linting can be triggered during the CI workflow.

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
N/A

#### Testing instructions:
Pre-requisite test changes
Add a `console.log('test');const test_var = false` line (with quite a few preceding tabs) into a `ts` and `svelte` boost file. Also add the same statement into a `tests/e2e/specs/` file.

Then:
- run `pnpm install` from the root of the Monorepo. Also, run the `jetpack install plugins/boost`.
- From the `Boost` folder, run `pnpm run lint` and observe the file being reformated by prettier and the errors found by eslint
- Repeat the pre-requisite test changes mentioned above and from the root of the Monorepo run the `pnpm run lint-required` command and notice the errors found by eslint.
- Repeat the pre-requisite test changes mentioned above (as well as other changes in JS files in Jetpack if you want) and from the root of the Monorepo run the `pnpm run reformat-files` command and notice that the files get re-formatted.
- Repeat the pre-requisite test changes mentioned above and try to do a test commit. The files should get reformatted and errors should occur from eslint which will stop the commit
- This PR should give a green check for JS linting
- Ensure that Boost is still working and that the assets can be built and watched for development.
